README.md
# ngx-treeview
An Angular Bootstrap treeview component
## Dependencies

- [Angular](https://angular.io)
- [Lodash](https://lodash.com)
- [Bootstrap 5.2](https://getbootstrap.com)

You can customize CSS yourself to break down dependencies to Bootstrap.

## Features

- Unlimited tree level
- State: disabled / collapse, expand

## Installation

After install the above dependencies, install `ngx-treeview` via:

```shell
npm i @zerot13/ngx-treeview --save
```

Once installed you need to import our main module in your application module:

```js
import { TreeviewModule } from '@zerot13/ngx-treeview';

@NgModule({
  declarations: [AppComponent, ...],
  imports: [TreeviewModule.forRoot(), ...],
  bootstrap: [AppComponent]
})
export class AppModule {
}
```

## Usage

#### Treeview:

```html
<ngx-treeview
  [config]="config"
  [items]="items"
  (selectedChange)="onSelectedChange($event)"
  (filterChange)="onFilterChange($event)"
>
</ngx-treeview>
```

#### Treeview with dropdown:

```html
<ngx-dropdown-treeview
  [buttonClass]="buttonClass"
  [config]="config"
  [items]="items"
  (selectedChange)="onSelectedChange($event)"
  (filterChange)="onFilterChange($event)"
>
</ngx-dropdown-treeview>
```

`config` is optional. This is the default configuration:

```js
{
   hasAllCheckBox: true,
   hasFilter: false,
   hasCollapseExpand: false,
   decoupleChildFromParent: false,
   maxHeight: 500
}
```

You can change default configuration easily because TreeviewConfig is injectable.

#### Pipe `ngxTreeview`:

To map your JSON objects to TreeItem objects.

```html
<ngx-dropdown-treeview
  [config]="config"
  [items]="items | ngxTreeview:'textField'"
  (selectedChange)="onSelectedChange($event)"
>
</ngx-dropdown-treeview>
```

#### Create a TreeviewItem:

```ts
const itCategory = new TreeviewItem(
{
  text: "Software",
  value: 9,
  children: [
    {
      text: "Programming",
      value: 91,
      children: [
        {
          text: "Frontend",
          value: 911,
          children: [
            { text: "Angular 12", value: 9112 },
            { text: "Angular 13", value: 9113 },
            { text: "Angular 14", value: 9114 },
            { text: "Angular 15", value: 9115, disabled: true },
            { text: "ReactJS", value: 9120 },
          ],
        },
        {
          text: "Backend",
          value: 912,
          children: [
            { text: "C#", value: 9121 },
            { text: "Java", value: 9122 },
            { text: "Python", value: 9123, checked: false },
          ],
        },
      ],
    },
  ],
},
{
  text: "Networking",
  value: 92,
  children: [
    { text: "Internet", value: 921 },
    { text: "Security", value: 922 },
    { text: "Switches", value: 923 },
  ],
});
```

You can pass the second paramater 'autoCorrectChecked' with value=true (default is false) in constructor of TreeviewItem to correct checked value of it and all of its descendants. In some cases, you need to push or pop children flexibly, checked of parent may be not correct. Then you need to call function correctChecked() to help to correct from root to its descendants.

```ts
const vegetableCategory = new TreeviewItem({
  text: "Vegetable",
  value: 2,
  children: [
    { text: "Salad", value: 21 },
    { text: "Potato", value: 22 },
  ],
});
vegetableCategory.children.push(
  new TreeviewItem({ text: "Mushroom", value: 23, checked: false })
);
vegetableCategory.correctChecked(); // need this to make 'Vegetable' node to change checked value from true to false
```

#### TreeviewEventParser:

Extract data from list of checked TreeviewItem and send it in parameter of event selectedChange. Some built-in TreeviewEventParser:

- DefaultTreeviewEventParser: return values of checked items.
- DownlineTreeviewEventParser: return list of checked items in orginal order with their ancestors.
- OrderDownlineTreeviewEventParser: return list of checked items in checked order with their ancestors. Note that: value of a leaf must be different from value of other leaves.

#### Templating:

See example 4 & 5.


CHANGELOG.md
# [14.0.0](https://www.npmjs.com/package/@treeview/ngx-treeview) (2023-06-15)

### Enhancement:

- Upgrade to Angular 14
- Upgrade to Bootstrap 5.2

src/app/book/book.component.ts
import { Component, OnInit } from '@angular/core';
import { TreeviewItem, TreeviewConfig } from 'ngx-treeview';
import { BookService } from './book.service';

@Component({
  selector: 'ngx-book',
  templateUrl: './book.component.html',
  providers: [
    BookService
  ]
})
export class BookComponent implements OnInit {
  dropdownEnabled = true;
  items: TreeviewItem[];
  values: number[];
  config = TreeviewConfig.create({
    hasAllCheckBox: true,
    hasFilter: true,
    hasCollapseExpand: true,
    decoupleChildFromParent: false,
    maxHeight: 400,
    compact: true
  });

  buttonClasses = [
    'btn-outline-primary',
    'btn-outline-secondary',
    'btn-outline-success',
    'btn-outline-danger',
    'btn-outline-warning',
    'btn-outline-info',
    'btn-outline-light',
    'btn-outline-dark'
  ];
  buttonClass = this.buttonClasses[0];

  constructor(
    private service: BookService
  ) { }

  ngOnInit(): void {
    this.items = this.service.getBooks();
  }

  onFilterChange(value: string): void {
    console.log('filter:', value);
  }
}


src/app/book/book.service.ts
import { TreeviewItem } from 'ngx-treeview';

export class BookService {
  getBooks(): TreeviewItem[] {
    const childrenCategory = new TreeviewItem({
      text: 'Children', value: 1, collapsed: true, children: [
        { text: 'Baby 3-5', value: 11 },
        { text: 'Baby 6-8', value: 12 },
        { text: 'Baby 9-12', value: 13 }
      ]
    });
    const itCategory = new TreeviewItem({
      text: 'IT', value: 9, children: [
        {
          text: 'Programming', value: 91, children: [{
            text: 'Frontend', value: 911, children: [
              { text: 'Angular 11', value: 9113 , disabled: true},
              { text: 'Angular 12', value: 9114, disabled: true },
              { text: 'Angular 13', value: 9115 },
              { text: 'Angular 14', value: 9116 },
              { text: 'ReactJS', value: 9117, disabled: true },
              { text: 'VueJS', value: 9118, disabled: true }
            ]
          }, 
          {
            text: 'Backend', value: 912, children: [
              { text: 'C#', value: 9121 },
              { text: 'Java', value: 9122 },
              { text: 'Python', value: 9123, checked: false, disabled: true },
              { text: 'Go', value: 9124, checked: false, disabled: true }
            ]
          }]
        },
        {
          text: 'Networking', value: 92, children: [
            { text: 'Internet', value: 921 },
            { text: 'Security', value: 922 },
            { text: 'Switches', value: 923 },
            { text: 'Routers', value: 924 },
          ]
        }
      ]
    });
    const teenCategory = new TreeviewItem({
      text: 'Teen', value: 2, collapsed: true, disabled: true, children: [
        { text: 'Adventure', value: 21 },
        { text: 'Science', value: 22 }
      ]
    });
    const extraCategory = new TreeviewItem({ text: 'Extra', value: 3, checked: false, });
    const othersCategory = new TreeviewItem({ text: 'Others', value: 3, checked: false, disabled: true });
    return [childrenCategory, itCategory, teenCategory, extraCategory,othersCategory];
  }
}
